File: apps/ios/Voiid/Voiid/Networking/SessionTeardown.swift
//
//  SessionTeardown.swift
//  Voiid
//
//  Removes this account's local footprint from the device.
//
//  WHY THIS EXISTS
//  ---------------
//  `AppSession.signOut()` clears the JWT and the profile UserDefaults key and nothing
//  else. Everything that actually holds the user's data survived it:
//
//   * the app-group SQLite store (`voiid.sqlite` + `-wal`/`-shm`) — and because
//     `ChatStore.loadConversations()` is local-first, the NEXT account to sign in on this
//     device rendered the previous user's conversations, messages and call history;
//   * `voiid_messages.json`, the plaintext decrypted-message blob;
//   * the E2E keychain (identity pickle, Olm session pickles, MLS state, and the backup
//     master secret);
//   * the registered VoIP push token, so the handset kept ringing for the account that
//     had signed out.
//
//  A polished, confirmed "Log Out" row sitting on top of that would make the leak more
//  prominent, not less. So log-out and teardown ship together, or neither ships.
//
//  ORDERING IS LOAD-BEARING
//  ------------------------
//  The auth token is NOT cleared here. Callers clear it afterwards (via
//  `AppSession.signOut()`), because step 1 below is an authenticated request and must go
//  out while the JWT still exists.
//
//  Every step is best-effort and independent: a failure in one must not prevent the rest.
//  Half a teardown is strictly worse than a slow one.
//

import Foundation

enum SessionTeardown {

    /// Wipe every local trace of the signed-in account. Call this BEFORE clearing the
    /// auth token — `AppSession.signOut()` immediately after is the intended sequence.
    ///
    /// Used by Log Out (`SettingsSheet`) and by Delete My Account (`EditProfileView`),
    /// which needs exactly the same local wipe after the server-side delete.
    @MainActor
    static func wipeLocalAccountState() async {
        await unregisterVoIPToken()   // 1 — needs the JWT, so it goes first
        wipeKeychain()                // 2
        wipeInMemoryState()           // 3 — BEFORE the files, or a flush rewrites them
        deleteLocalFiles()            // 4
        reopenEmptyDatabase()         // 5
        resetLegacyImportFlag()       // 6
        clearCaches()                 // 7
        MissedCallNotifier.cancelAll()// 8 — daemon-held banners outlive the process
        // 9 — the caller clears TokenStore / calls auth.logout().
    }

    // MARK: - 3. In-memory state
    //
    // Deleting files is only half a teardown: signing out does NOT restart the process, so
    // every in-memory copy of the previous account's data survives into the next session.
    // This step must run BEFORE deleteLocalFiles(), because any of these holders could
    // otherwise flush its stale contents back to disk after the delete.

    @MainActor
    private static func wipeInMemoryState() {
        // Plaintext messages and live Olm sessions.
        ChatEngine.shared.wipeInMemoryState()
        // Contact names, phone numbers and photos from the previous user's address book.
        UserDirectory.shared.wipe()
        // The once-per-process bootstrap latch and the in-memory identity: without this the
        // next account's bootstrap() returns immediately and reuses the old identity.
        E2EManager.shared.resetForSignOut()
        // Close the database BEFORE unlinking its files — see closeForTeardown().
        VoiidDatabase.shared.closeForTeardown()
    }

    // MARK: - 5. Fresh database

    /// The process keeps running, so hand the next sign-in a working, empty database rather
    /// than leaving every local read disabled for the rest of the session.
    @MainActor
    private static func reopenEmptyDatabase() {
        VoiidDatabase.shared.reopenAfterTeardown()
    }

    // MARK: - 1. VoIP token

    /// Tell the backend to stop sending VoIP pushes to this device. The route explicitly
    /// accepts a null token as "unregister".
    ///
    /// Failure is ignored: we are signing out either way, and blocking a log-out on a
    /// network round-trip would be a worse bug than a stale push token. Worst case the
    /// next sign-in on this device overwrites the row.
    @MainActor
    private static func unregisterVoIPToken() async {
        guard TokenStore.shared.jwt != nil,
              let deviceId = E2EManager.shared.deviceId else { return }
        struct Body: Encodable {
            let device_id: String
            let voip_token: String?
        }
        _ = try? await APIClient().request("POST", "devices/voip-token",
                                           body: Body(device_id: deviceId, voip_token: nil),
                                           as: EmptyResponse.self)
    }

    // MARK: - 2. Keychain

    /// The iOS keychain survives app uninstall, so leaving these behind means a reinstall
    /// (or a different account on the same handset) resurrects a stale identity.
    ///
    /// `clearMasterSecret()` is called explicitly as well as wiping `com.voiid.e2e`: it is
    /// the one item whose loss the user was warned about in the confirmation dialog, and
    /// stating it here keeps that promise findable.
    @MainActor
    private static func wipeKeychain() {
        E2EManager.shared.clearMasterSecret()
        for service in ["com.voiid.e2e", "com.voiid.sessions", "com.voiid.mls"] {
            KeychainData(service: service).wipeService()
        }
    }

    // MARK: - 3. Local files

    /// The database and the decrypted-message blob, in both their app-group location and
    /// the Application Support fallback an unentitled build would have used.
    private static func deleteLocalFiles() {
        let fm = FileManager.default
        var targets: [URL] = []

        let appSupport = fm.urls(for: .applicationSupportDirectory, in: .userDomainMask).first

        for base in [AppGroup.containerURL, appSupport].compactMap({ $0 }) {
            let db = base.appendingPathComponent("voiid.sqlite")
            targets.append(db)
            targets.append(URL(fileURLWithPath: db.path + "-wal"))
            targets.append(URL(fileURLWithPath: db.path + "-shm"))
            targets.append(base.appendingPathComponent("voiid_messages.json"))
            // Phase 2: the per-conversation message shards live under messages/.
            targets.append(base.appendingPathComponent("messages", isDirectory: true))
        }

        for url in targets {
            try? fm.removeItem(at: url)   // removeItem deletes a directory recursively
        }
    }

    // MARK: - 4. Legacy import flag

    /// `LocalStore.importLegacyMessageBlobIfNeeded()` is gated on this flag and runs
    /// exactly once per install. Leaving it set means a future account on this device
    /// silently never runs the legacy import.
    private static func resetLegacyImportFlag() {
        // Mirrors LocalStore.importFlagKey, which is private to that type.
        UserDefaults.standard.removeObject(forKey: "voiid.localstore.imported_message_blob_v1")
    }

    // MARK: - 5. Caches

    /// In-memory decrypted media and the URL cache both hold the previous account's
    /// content until the process dies. Sign-out does not restart the process.
    @MainActor
    private static func clearCaches() {
        URLCache.shared.removeAllCachedResponses()
        MediaCache.shared.clear()
        AvatarCache.clear()   // previous account's faces must not survive into the next login
    }
}

File: apps/android/app/src/main/java/com/voiid/app/net/SessionTeardown.kt
package com.voiid.app.net

import android.content.Context
import com.voiid.app.store.UserDirectory
import com.voiid.app.store.VoiidDatabase

/**
 * Removes this account's local footprint from the device. Port of iOS
 * `SessionTeardown.swift`.
 *
 * WHY THIS EXISTS
 * ---------------
 * `AppSession.signOut()` cleared only the JWT (`TokenStore.clear()` via
 * `AuthService.logout()`). Everything that actually holds the user's data survived it:
 *
 *  - the Room database (`voiid.db`) — and because every read here is local-first (see
 *    `VoiidDatabase`'s class doc), the NEXT account signed in on this device rendered the
 *    previous user's conversations, messages and call history;
 *  - `voiid_messages.json`, ChatEngine's plaintext decrypted-message store, and its
 *    in-memory Olm sessions;
 *  - the encrypted E2E prefs (identity pickle, pickle key, device id, prekey counters) and
 *    the `bootstrapped` latch, so the next account's `bootstrap()` silently reused the old
 *    identity;
 *  - the in-memory contact-name mirror in `UserDirectory`.
 *
 * ORDERING IS LOAD-BEARING — mirrors iOS exactly: wipe in-memory state BEFORE deleting
 * files/tables, or a holder still running (a send, a poll) could flush its stale contents
 * right back to disk after the delete. Every step is best-effort and independent: a
 * failure in one must not prevent the rest.
 */
object SessionTeardown {

    /**
     * Wipe every local trace of the signed-in account. Call this BEFORE clearing the auth
     * token — `AppSession.signOut()` calling `auth.logout()` immediately after is the
     * intended sequence.
     */
    fun wipeLocalAccountState(context: Context) {
        val appContext = context.applicationContext

        // 1 — in-memory state, BEFORE the files/tables, or a flush rewrites them.
        runCatching { ChatEngine.get(appContext).wipeInMemoryState() }
        runCatching { UserDirectory.wipe() }
        runCatching { E2EManager.get(appContext).resetForSignOut() }
        runCatching { GroupEngine.get(appContext).resetForSignOut() }

        // 2 — the local-first store every screen reads from.
        runCatching { VoiidDatabase.wipeAllForSignOut(appContext) }

        // 3 — ChatEngine's own plaintext file store (separate from Room; see its class doc).
        runCatching { java.io.File(appContext.filesDir, "voiid_messages.json").delete() }
        runCatching { java.io.File(appContext.filesDir, "voiid_messages.json.tmp").delete() }
        // Phase 2: the per-conversation message shards.
        runCatching { java.io.File(appContext.filesDir, "messages").deleteRecursively() }

        // 4 — decrypted media cache (memory + disk), or the previous account's photos/voice
        // notes stay readable on this device behind the login screen.
        runCatching { com.voiid.app.main.MediaCache.clear(appContext) }
    }
}
